### PR TITLE
HW8. Frontend routing, loaders, fetching data 

### PR DESCRIPTION
**Pull request by izzipizzi**
Last commit: https://github.com/izzipizzi/test-project/pull/8/commits/91ce199e3c46864a067c9b70b751da40fcf78a12

image

Toast After post creation.
image

Loading single post:
image

### CODE REVIEW

**Comment by MMaksymivUa** [comment]
Too many changes. You should split for several PR's: updating project structure(server/client folder), adding backend routing, adding frontend routing

File: shared/interfaces/Tag.d.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+    text: string;
+}
+export interface PostTagCreateData {
+    postId: string;
+    text: string;
+}
+export interface PostTagDeleteData {
+    postId: string;
+    tagId: string;
+}
```
Comment (MMaksymivUa) [inline shared/interfaces/Tag.d.ts:12]: Please follow one convention of naming interface's for all: with -Model suffix; with I- prefix, with -DTO suffix. Do not mix them

File: src/app/app-routing.module.ts [hunk 2]
```diff
@@ -1,20 +1,29 @@
...
   {
-    path: '', component: PostsComponent
-  }
+    path: 'auth',
+    loadChildren: () => import('./auth/auth.module').then((m) => m.AuthModule),
+  },
+  {
+    path: '',
+    loadChildren: () => import('./application/application.module').then((m) => m.ApplicationModule),
+  },
```
Comment (MMaksymivUa) [inline src/app/app-routing.module.ts:17]: no need to do that. You have a bootstrap AppModule, no need to lazy-load another

File: src/app/app.module.ts [hunk 3]
```diff
@@ -3,27 +3,14 @@ import { BrowserModule } from '@angular/platform-browser';
...
-    CommonModule,
-    BrowserModule,
-    SharedModule,
-    AppRoutingModule,
-    FormsModule,
-    ReactiveFormsModule,
-    BrowserAnimationsModule,
-  ],
+  declarations: [AppComponent],
+  imports: [CommonModule, SharedModule, BrowserModule, BrowserAnimationsModule, AppRoutingModule, HttpClientModule],
```
Comment (MMaksymivUa) [inline src/app/app.module.ts:13]: No need to import common and browser module, only browser for bootstrap module

File: src/app/application/add-post/add-post.component.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+  addPost(postData: CreatePostData) {
+    this.isLoading.next(true);
+    this.postsService.addPost(postData).subscribe((data) => {
+      this.isLoading.next(false);
+      this.toastService.notification$.next(data.message);
+      this.router.navigate(['/']);
+    });
+  }
+}
```
Comment (MMaksymivUa) [inline src/app/application/add-post/add-post.component.ts:32]: Create special spinner service for that

File: src/app/application/application-routing.module.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { NgModule } from '@angular/core';
```
Comment (MMaksymivUa) [inline src/app/application/application-routing.module.ts:1]: maybe it is a postmodule?

File: src/app/application/posts/post-info/post-info.component.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  ngOnInit(): void {
+    this.route.params.pipe(switchMap((params) => this.postsService.getPostById(params['postId']))).subscribe((post) => {
+      this.post = post;
+      this.isLoading.next(false);
+    });
+  }
+  handleLikeClick(postId: string) {
+    //here will be added post like feature, after authorization implementation
+  }
+}
```
Comment (MMaksymivUa) [inline src/app/application/posts/post-info/post-info.component.ts:27]: do not create a rxjs stream for each component. Also do not forget about unsubscribe. Create shared spinnerservice

File: src/app/application/posts/posts-routing.module.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    path: '',
+    component: PostsComponent,
+  },
+];
+
+@NgModule({
+  imports: [RouterModule.forChild(routes)],
+  exports: [RouterModule],
+})
+export class PostsRoutingModule {}
```
Comment (MMaksymivUa) [inline src/app/application/posts/posts-routing.module.ts:16]: rename application module to post module, delete this one

File: src/app/shared/components/post/post.component.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+  isOpened: boolean = false;
+
+  constructor() {}
+
+  ngOnInit(): void {}
+
+  toggleOpened(event: MouseEvent): void {
+    event.stopPropagation();
+    this.isOpened ? (this.isOpened = false) : (this.isOpened = true);
```
Comment (MMaksymivUa) [inline src/app/shared/components/post/post.component.ts:22]: this.isOpened = !this.isOpened

File: src/app/shared/components/post/post.component.ts [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  get postLikes(): string {
+    if (this.post?.likes.length) {
+      return this.post.likes.length > 0 ? 'warn' : '';
+    }
+    return '';
+  }
```
Comment (MMaksymivUa) [inline src/app/shared/components/post/post.component.ts:30]:
```  
get postLikes(): string {
      return this.post?.likes?.length ? 'warn' : '';
  }
```
keep it simple

File: src/app/shared/components/post/post.component.ts [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+  addPostLike(event: MouseEvent) {
+    event.stopPropagation();
```
Comment (MMaksymivUa) [inline src/app/shared/components/post/post.component.ts:33]: why you stop propagation for each click?

**Comment by MMaksymivUa** [comment]
Good job. Fix comments please

File: src/app/shared/components/post/post.component.ts (hunk 10 above)
Comment (izzipizzi) [reply src/app/shared/components/post/post.component.ts:33]: When I click on card with post, i navigate to page with single post. I need stop propagation to not trigger navigation, and add like to post.

File: shared/interfaces/Tag.d.ts (hunk 1 above)
Comment (izzipizzi) [reply shared/interfaces/Tag.d.ts:12]: I fixed this, according Anton Fandul remarks.

**Pull request merged**